ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Decision.hs

```diff
@@ -0,0 +1,345 @@
+import Data.Map.Strict qualified as Map
+import Data.Sequence.Strict (StrictSeq)
+import Data.Sequence.Strict qualified as StrictSeq
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound.V2.State
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound.V2.Types
+import Ouroboros.Network.Protocol.ObjectDiffusion.Type
+import System.Random (StdGen)
+import System.Random.Shuffle (shuffle')
+
+strictSeqToSet :: Ord a => StrictSeq a -> Set a
+strictSeqToSet = Set.fromList . Foldable.toList
+
+-- | Make download decisions.
+makeDecisions ::
+  forall peerAddr objectId object.
+  ( Ord peerAddr
+  , Ord objectId
+  ) =>
+  DecisionContext peerAddr objectId object ->
+  -- | New decisions
+  Map peerAddr (PeerDecision objectId object)
+makeDecisions DecisionContext{dcRng, dcHasObject, dcDecisionPolicy, dcGlobalState, dcPrevDecisions} =
+  let
+    -- A subset of peers are currently executing a decision. We shouldn't update the decision for them
+    frozenPeersToDecisions = Map.filter (\PeerDecision{pdStatus} -> pdStatus == DecisionBeingActedUpon) dcPrevDecisions
+
+    -- We do it in two steps, because computing the acknowledgment tell which objects from dpsObjectsAvailableIds sets of each peer won't actually be available anymore (as soon as we ack them),
+    -- so that the pickObjectsToReq function can take this into account.
+    (ackAndRequestIdsDecisions, peerToIdsToAck) = computeAck dcHasObject dcDecisionPolicy dcGlobalState frozenPeersToDecisions
+    peersToObjectsToReq =
+      pickObjectsToReq
+        dcRng
+        dcHasObject
+        dcDecisionPolicy
+        dcGlobalState
+        frozenPeersToDecisions
+        peerToIdsToAck
+   in
+    Map.intersectionWith
+      (\decision objectsToReqIds -> decision{pdObjectsToReqIds = objectsToReqIds})
+      ackAndRequestIdsDecisions
+      peersToObjectsToReq
+
+-- | The ids to ack are the longest prefix of outstandingFifo of each peer that match the following criteria:
+-- * either the object is owt pool for the peer who has downloaded it
+-- * or the object is already in pool
+computeAck ::
+  forall peerAddr objectId object.
+  ( Ord peerAddr
+  , Ord objectId
+  ) =>
+  (objectId -> Bool) ->
+  DecisionPolicy ->
+  DecisionGlobalState peerAddr objectId object ->
+  -- | Frozen peers and their previous decisions
+  Map peerAddr (PeerDecision objectId object) ->
+  ( Map peerAddr (PeerDecision objectId object)
+  , Map peerAddr (Set objectId)
+  )
+computeAck poolHasObject DecisionPolicy{dpMaxNumObjectIdsReq, dpMaxNumObjectsOutstanding} DecisionGlobalState{dgsPeerStates} frozenPeersToDecisions =
+  let
+    -- We shouldn't create a new decision for peers that are currently executing a decision
+    filteredPeerStates = Map.withoutKeys dgsPeerStates (Map.keysSet frozenPeersToDecisions)
+    (decisions, peerToIdsToAck) =
+      Map.foldlWithKey' computeAckForPeer (Map.empty, Map.empty) filteredPeerStates
```

Comment:
It seems like you're unconditionally inserting each input peer processed by this fold into the accumulator map, which makes me think that this could be defined more precisely via [Map.mapAccumWithKey](https://hackage.haskell.org/package/containers-0.4.0.0/docs/Data-Map.html#v:mapAccumWithKey). 

I think such an implementation _should_ be more perfomant (single traversal (O(n)) vs. having to manually insert peers into the accumulator (O(n.log(n))) _and_ have the added benefit of preserving `length (fmap computeAckForPeer filteredPeerStates) == length filteredPeerStates` via parametricity.

Would this be possible?

Reply:
At first glance, I think we should be able to do that.

Reply:
I think I made it even simpler than that, with `Map.map`. Let me know what you think